File: src/robot_life/behavior/resources.py
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from time import monotonic

from robot_life.common.schemas import ResourceGrant, new_id

# Standard resource names in the system
RESOURCE_NAMES = [
    "AudioOut",
    "HeadMotion",
    "BodyMotion",
    "FaceExpression",
    "AttentionTarget",
    "DialogContext",
]


class ResourceMode(str, Enum):
    """How a resource can be shared between behaviors."""

    EXCLUSIVE = "EXCLUSIVE"
    SHARED = "SHARED"
    DUCKING = "DUCKING"


@dataclass
class ResourceDef:
    """Definition of a resource's properties."""

    name: str
    mode: ResourceMode
    priority_level: int  # Higher = more important


@dataclass
class ResourceOwner:
    grant_id: str
    behavior_id: str
    priority: int
    end_time: float


# Standard resource definitions
RESOURCE_DEFS = {
    "AudioOut": ResourceDef("AudioOut", ResourceMode.EXCLUSIVE, 3),
    "HeadMotion": ResourceDef("HeadMotion", ResourceMode.SHARED, 2),
    "BodyMotion": ResourceDef("BodyMotion", ResourceMode.SHARED, 2),
    "FaceExpression": ResourceDef("FaceExpression", ResourceMode.EXCLUSIVE, 2),
    "AttentionTarget": ResourceDef("AttentionTarget", ResourceMode.SHARED, 1),
    "DialogContext": ResourceDef("DialogContext", ResourceMode.DUCKING, 2),
}


class ResourceManager:
    """
    Manages resource allocation and ownership.

    Enforces exclusive access, sharing, ducking, and degradation rules.
    """

    def __init__(self) -> None:
        # resource_name -> list[ResourceOwner]
        self._owners: dict[str, list[ResourceOwner]] = {}
        self._active_grants: dict[str, ResourceGrant] = {}
        self._stats: dict[str, int] = {
            "grant_requests": 0,
            "grant_granted": 0,
            "grant_denied": 0,
            "preemptions": 0,
        }

    def request_grant(
        self,
        trace_id: str,
        decision_id: str,
        behavior_id: str,
        required_resources: list[str],
        optional_resources: list[str],
        priority: int = 2,
        duration_ms: int = 5000,
    ) -> ResourceGrant:
        now = monotonic()
        end_time = now + (duration_ms / 1000.0)
        self._cleanup_expired(now)
        self._stats["grant_requests"] += 1

        grant_id = new_id()
        granted: list[str] = []
        denied: list[str] = []
        reason_parts: list[str] = []
        preempt_plan: dict[str, list[str]] = {}

        # Phase 1: verify all required resources can be allocated.
        for resource_name in required_resources:
            can_allocate, preempt_grants = self._evaluate_allocation(resource_name, priority, now)
            if not can_allocate:
                denied.append(resource_name)
                reason_parts.append(self._build_conflict_reason(resource_name, now))
                continue
            preempt_plan[resource_name] = preempt_grants

        if denied:
            self._stats["grant_denied"] += 1
            grant = ResourceGrant(
                grant_id=grant_id,
                trace_id=trace_id,
                decision_id=decision_id,
                granted=False,
                granted_resources=[],
                denied_resources=denied,
                reason=f"required_resources_denied: {', '.join(reason_parts)}",
            )
            # NOTE: Denied grants are NOT stored in _active_grants
            # to prevent unbounded memory growth from failed requests.
            return grant

        # Phase 2: allocate required resources.
        for resource_name in required_resources:
            self._allocate(
                resource_name=resource_name,
                owner=ResourceOwner(
                    grant_id=grant_id,
                    behavior_id=behavior_id,
                    priority=priority,
                    end_time=end_time,
                ),
                preempt_grants=preempt_plan.get(resource_name, []),
            )
            granted.append(resource_name)

        # Phase 3: best-effort optional resource allocation.
        for resource_name in optional_resources:
            can_allocate, preempt_grants = self._evaluate_allocation(resource_name, priority, now)
            if not can_allocate:
                continue
            self._allocate(
                resource_name=resource_name,
                owner=ResourceOwner(
                    grant_id=grant_id,
                    behavior_id=behavior_id,
                    priority=priority,
                    end_time=end_time,
                ),
                preempt_grants=preempt_grants,
            )
            granted.append(resource_name)

        grant = ResourceGrant(
            grant_id=grant_id,
            trace_id=trace_id,
            decision_id=decision_id,
            granted=True,
            granted_resources=granted,
            denied_resources=[],
            reason=f"granted_{len(granted)}_resources" if granted else "no_resources_needed",
        )
        self._stats["grant_granted"] += 1
        self._active_grants[grant_id] = grant
        return grant

    def _evaluate_allocation(
        self,
        resource_name: str,
        priority: int,
        now: float,
    ) -> tuple[bool, list[str]]:
        owners = self._owners.get(resource_name, [])
        owners = [owner for owner in owners if owner.end_time > now]
        self._owners[resource_name] = owners

        if not owners:
            return True, []

        resource_def = RESOURCE_DEFS.get(resource_name)
        if resource_def is None:
            return False, []

        if resource_def.mode == ResourceMode.EXCLUSIVE:
            highest_owner_priority = max(owner.priority for owner in owners)
            if priority > highest_owner_priority:
                return True, [owner.grant_id for owner in owners]
            return False, []

        # SHARED/DUCKING are both multi-owner.
        return True, []

    def _allocate(self, resource_name: str, owner: ResourceOwner, preempt_grants: list[str]) -> None:
        if preempt_grants:
            self._stats["preemptions"] += len(preempt_grants)
            for preempt_grant_id in preempt_grants:
                self._remove_grant_from_resource(resource_name, preempt_grant_id)

        owners = self._owners.setdefault(resource_name, [])
        if not any(existing.grant_id == owner.grant_id for existing in owners):
            owners.append(owner)

    def _remove_grant_from_resource(self, resource_name: str, grant_id: str) -> None:
        owners = self._owners.get(resource_name, [])
        if not owners:
            return

        new_owners = [owner for owner in owners if owner.grant_id != grant_id]
        if new_owners:
            self._owners[resource_name] = new_owners
        else:
            self._owners.pop(resource_name, None)

        grant = self._active_grants.get(grant_id)
        if grant is not None:
            grant.granted_resources = [res for res in grant.granted_resources if res != resource_name]

    def _cleanup_expired(self, now: float) -> None:
        for resource_name in list(self._owners.keys()):
            owners = self._owners[resource_name]
            expired_grant_ids = [owner.grant_id for owner in owners if owner.end_time <= now]
            alive = [owner for owner in owners if owner.end_time > now]
            if alive:
                self._owners[resource_name] = alive
            else:
                self._owners.pop(resource_name, None)

            for grant_id in expired_grant_ids:
                grant = self._active_grants.get(grant_id)
                if grant is not None:
                    grant.granted_resources = [
                        resource for resource in grant.granted_resources if resource != resource_name
                    ]

    def _build_conflict_reason(self, resource_name: str, now: float) -> str:
        owners = self._owners.get(resource_name, [])
        active_owners = [owner for owner in owners if owner.end_time > now]
        if not active_owners:
            return f"{resource_name}:conflict"

        labels = ",".join(
            f"{owner.behavior_id}({self._priority_label(owner.priority)})" for owner in active_owners
        )
        return f"{resource_name}:owned_by_{labels}"

    def release_grant(self, grant_id: str) -> None:
        grant = self._active_grants.pop(grant_id, None)
        if grant is None:
            return

        for resource_name in list(self._owners.keys()):
            self._remove_grant_from_resource(resource_name, grant_id)

    def force_release_all(self) -> None:
        """Emergency path: release every active grant immediately."""
        for grant_id in list(self._active_grants.keys()):
            self.release_grant(grant_id)

    def get_resource_status(self) -> dict[str, str]:
        now = monotonic()
        self._cleanup_expired(now)
        status: dict[str, str] = {}

        for resource_name in RESOURCE_DEFS:
            owners = self._owners.get(resource_name, [])
            if not owners:
                status[resource_name] = "free"
                continue

            owner_parts = []
            for owner in sorted(owners, key=lambda item: item.priority, reverse=True):
                ttl_ms = max(0, int((owner.end_time - now) * 1000))
                owner_parts.append(
                    f"{owner.behavior_id}({self._priority_label(owner.priority)},{ttl_ms}ms)"
                )

            mode = RESOURCE_DEFS[resource_name].mode
            prefix = "owned_by" if mode == ResourceMode.EXCLUSIVE else "shared_by"
            status[resource_name] = f"{prefix}_{'|'.join(owner_parts)}"

        return status

    def debug_snapshot(self) -> dict:
        now = monotonic()
        self._cleanup_expired(now)
        owners_view: dict[str, list[dict[str, str | int]]] = {}
        for resource_name in RESOURCE_DEFS:
            owners = self._owners.get(resource_name, [])
            serialized = []
            for owner in sorted(owners, key=lambda item: item.priority, reverse=True):
                serialized.append(
                    {
                        "grant_id": owner.grant_id,
                        "behavior_id": owner.behavior_id,
                        "priority": self._priority_label(owner.priority),
                        "ttl_ms": max(0, int((owner.end_time - now) * 1000)),
                    }
                )
            owners_view[resource_name] = serialized

        return {
            "stats": dict(self._stats),
            "active_grants": len(self._active_grants),
            "status": self.get_resource_status(),
            "owners": owners_view,
        }

    @staticmethod
    def _priority_label(internal_priority: int) -> str:
        if 0 <= internal_priority <= 3:
            return f"P{3 - internal_priority}"
        return f"internal_{internal_priority}"

File: Robot_Life_CPP/src/behavior/resources.cpp
#include "robot_life_cpp/behavior/resources.hpp"

#include <algorithm>
#include <limits>
#include <sstream>

namespace robot_life_cpp::behavior {

ResourceManager::ResourceManager() {
  resource_defs_.emplace("camera", ResourceDef{"camera", ResourceMode::Shared, 2});
  resource_defs_.emplace("audio", ResourceDef{"audio", ResourceMode::Exclusive, 2});
  resource_defs_.emplace("gpu", ResourceDef{"gpu", ResourceMode::Shared, 2});
  resource_defs_.emplace("AudioOut", ResourceDef{"AudioOut", ResourceMode::Exclusive, 3});
  resource_defs_.emplace("HeadMotion", ResourceDef{"HeadMotion", ResourceMode::Shared, 2});
  resource_defs_.emplace("BodyMotion", ResourceDef{"BodyMotion", ResourceMode::Shared, 2});
  resource_defs_.emplace("FaceExpression", ResourceDef{"FaceExpression", ResourceMode::Exclusive, 2});
  resource_defs_.emplace("AttentionTarget", ResourceDef{"AttentionTarget", ResourceMode::Shared, 1});
  resource_defs_.emplace("DialogContext", ResourceDef{"DialogContext", ResourceMode::Ducking, 2});
}

ResourceGrant ResourceManager::request_grant(
    const std::string& trace_id,
    const std::string& decision_id,
    const std::string& behavior_id,
    const std::vector<std::string>& required_resources,
    const std::vector<std::string>& optional_resources,
    int priority,
    int duration_ms,
    double now_mono_s) {
  cleanup_expired(now_mono_s);
  const auto end_time = now_mono_s + (static_cast<double>(std::max(1, duration_ms)) / 1000.0);

  const auto grant_id = common::new_id();
  std::vector<std::string> granted{};
  std::vector<std::string> denied{};
  std::vector<std::string> reasons{};
  std::unordered_map<std::string, std::vector<std::string>> preempt_plan{};

  for (const auto& resource : required_resources) {
    std::vector<std::string> preempt_grants{};
    if (!evaluate_allocation(resource, priority, now_mono_s, &preempt_grants)) {
      denied.push_back(resource);
      reasons.push_back(build_conflict_reason(resource, now_mono_s));
      continue;
    }
    preempt_plan.emplace(resource, std::move(preempt_grants));
  }

  if (!denied.empty()) {
    ResourceGrant denied_grant{};
    denied_grant.grant_id = grant_id;
    denied_grant.trace_id = trace_id;
    denied_grant.decision_id = decision_id;
    denied_grant.granted = false;
    denied_grant.denied_resources = std::move(denied);
    std::ostringstream oss;
    oss << "required_resources_denied";
    for (const auto& r : reasons) {
      oss << ":" << r;
    }
    denied_grant.reason = oss.str();
    return denied_grant;
  }

  for (const auto& resource : required_resources) {
    const auto it = preempt_plan.find(resource);
    const std::vector<std::string> preempt_grants = it == preempt_plan.end()
                                                        ? std::vector<std::string>{}
                                                        : it->second;
    allocate(
        resource,
        ResourceOwner{
            .grant_id = grant_id,
            .behavior_id = behavior_id,
            .priority = priority,
            .end_time = end_time,
        },
        preempt_grants);
    granted.push_back(resource);
  }

  for (const auto& resource : optional_resources) {
    std::vector<std::string> preempt_grants{};
    if (!evaluate_allocation(resource, priority, now_mono_s, &preempt_grants)) {
      continue;
    }
    allocate(
        resource,
        ResourceOwner{
            .grant_id = grant_id,
            .behavior_id = behavior_id,
            .priority = priority,
            .end_time = end_time,
        },
        preempt_grants);
    granted.push_back(resource);
  }

  ResourceGrant ok{};
  ok.grant_id = grant_id;
  ok.trace_id = trace_id;
  ok.decision_id = decision_id;
  ok.granted = true;
  ok.granted_resources = std::move(granted);
  ok.reason = ok.granted_resources.empty()
                  ? "no_resources_needed"
                  : "granted_" + std::to_string(ok.granted_resources.size()) + "_resources";
  active_grants_[grant_id] = ok;
  return ok;
}

void ResourceManager::release_grant(const std::string& grant_id) {
  auto grant_it = active_grants_.find(grant_id);
  if (grant_it == active_grants_.end()) {
    return;
  }
  for (auto it = owners_.begin(); it != owners_.end();) {
    auto& owners = it->second;
    owners.erase(
        std::remove_if(
            owners.begin(),
            owners.end(),
            [&](const ResourceOwner& owner) { return owner.grant_id == grant_id; }),
        owners.end());
    if (owners.empty()) {
      it = owners_.erase(it);
    } else {
      ++it;
    }
  }
  active_grants_.erase(grant_it);
}

void ResourceManager::force_release_all() {
  std::vector<std::string> grant_ids{};
  grant_ids.reserve(active_grants_.size());
  for (const auto& [grant_id, _] : active_grants_) {
    grant_ids.push_back(grant_id);
  }
  for (const auto& grant_id : grant_ids) {
    release_grant(grant_id);
  }
}

std::unordered_map<std::string, std::string> ResourceManager::get_resource_status(double now_mono_s) {
  cleanup_expired(now_mono_s);
  std::unordered_map<std::string, std::string> status{};
  for (const auto& [resource_name, def] : resource_defs_) {
    auto owners_it = owners_.find(resource_name);
    if (owners_it == owners_.end() || owners_it->second.empty()) {
      status[resource_name] = "free";
      continue;
    }
    std::vector<ResourceOwner> owners = owners_it->second;
    std::sort(
        owners.begin(),
        owners.end(),
        [](const ResourceOwner& lhs, const ResourceOwner& rhs) { return lhs.priority > rhs.priority; });
    std::ostringstream oss;
    oss << (def.mode == ResourceMode::Exclusive ? "owned_by_" : "shared_by_");
    bool first = true;
    for (const auto& owner : owners) {
      if (!first) {
        oss << "|";
      }
      first = false;
      const auto ttl_ms = std::max(0, static_cast<int>((owner.end_time - now_mono_s) * 1000.0));
      oss << owner.behavior_id << "(" << priority_label(owner.priority) << "," << ttl_ms << "ms)";
    }
    status[resource_name] = oss.str();
  }
  return status;
}

std::string ResourceManager::priority_label(int internal_priority) {
  if (internal_priority < 0 || internal_priority > 3) {
    return "internal_" + std::to_string(internal_priority);
  }
  return "P" + std::to_string(3 - internal_priority);
}

bool ResourceManager::evaluate_allocation(
    const std::string& resource_name,
    int priority,
    double now_mono_s,
    std::vector<std::string>* preempt_grants) {
  cleanup_expired(now_mono_s);
  const auto def_it = resource_defs_.find(resource_name);
  if (def_it == resource_defs_.end()) {
    return false;
  }
  auto owners_it = owners_.find(resource_name);
  if (owners_it == owners_.end() || owners_it->second.empty()) {
    return true;
  }

  auto& owners = owners_it->second;
  if (def_it->second.mode != ResourceMode::Exclusive) {
    return true;
  }

  int highest_owner_priority = std::numeric_limits<int>::min();
  for (const auto& owner : owners) {
    highest_owner_priority = std::max(highest_owner_priority, owner.priority);
  }
  if (priority <= highest_owner_priority) {
    return false;
  }
  if (preempt_grants != nullptr) {
    preempt_grants->clear();
    preempt_grants->reserve(owners.size());
    for (const auto& owner : owners) {
      preempt_grants->push_back(owner.grant_id);
    }
  }
  return true;
}

void ResourceManager::allocate(
    const std::string& resource_name,
    const ResourceOwner& owner,
    const std::vector<std::string>& preempt_grants) {
  for (const auto& preempt_id : preempt_grants) {
    remove_grant_from_resource(resource_name, preempt_id);
  }
  auto& owners = owners_[resource_name];
  const auto exists = std::find_if(
      owners.begin(),
      owners.end(),
      [&](const ResourceOwner& item) { return item.grant_id == owner.grant_id; });
  if (exists == owners.end()) {
    owners.push_back(owner);
  }
}

void ResourceManager::remove_grant_from_resource(
    const std::string& resource_name,
    const std::string& grant_id) {
  auto owners_it = owners_.find(resource_name);
  if (owners_it == owners_.end()) {
    return;
  }
  auto& owners = owners_it->second;
  owners.erase(
      std::remove_if(
          owners.begin(),
          owners.end(),
          [&](const ResourceOwner& owner) { return owner.grant_id == grant_id; }),
      owners.end());
  if (owners.empty()) {
    owners_.erase(owners_it);
  }

  auto grant_it = active_grants_.find(grant_id);
  if (grant_it != active_grants_.end()) {
    auto& granted_resources = grant_it->second.granted_resources;
    granted_resources.erase(
        std::remove(granted_resources.begin(), granted_resources.end(), resource_name),
        granted_resources.end());
    if (granted_resources.empty()) {
      active_grants_.erase(grant_it);
    }
  }
}

void ResourceManager::cleanup_expired(double now_mono_s) {
  for (auto it = owners_.begin(); it != owners_.end();) {
    auto& owners = it->second;
    owners.erase(
        std::remove_if(
            owners.begin(),
            owners.end(),
            [&](const ResourceOwner& owner) { return owner.end_time <= now_mono_s; }),
        owners.end());
    if (owners.empty()) {
      it = owners_.erase(it);
    } else {
      ++it;
    }
  }
}

std::string ResourceManager::build_conflict_reason(
    const std::string& resource_name,
    double now_mono_s) const {
  auto owners_it = owners_.find(resource_name);
  if (owners_it == owners_.end() || owners_it->second.empty()) {
    return resource_name + ":conflict";
  }
  std::ostringstream oss;
  oss << resource_name << ":owned_by_";
  bool first = true;
  for (const auto& owner : owners_it->second) {
    if (owner.end_time <= now_mono_s) {
      continue;
    }
    if (!first) {
      oss << ",";
    }
    first = false;
    oss << owner.behavior_id << "(" << priority_label(owner.priority) << ")";
  }
  return oss.str();
}

}  // namespace robot_life_cpp::behavior
